nomos-services/cryptarchia-consensus/src/mix/adapters/libp2p.rs

```diff
@@ -0,0 +1,86 @@
+use nomos_mix_service::{
+    backends::libp2p::{
+        Libp2pNetworkBackend, Libp2pNetworkBackendEvent, Libp2pNetworkBackendEventKind,
+        Libp2pNetworkBackendMessage,
+    },
+    NetworkMsg, NetworkService,
+};
+use overwatch_rs::services::{relay::OutboundRelay, ServiceData};
+use tokio::sync::broadcast::{self, error::RecvError};
+use tokio_stream::{wrappers::BroadcastStream, StreamExt};
+
+use crate::mix::{BoxedStream, NetworkAdapter};
+
+const BUFFER_SIZE: usize = 64;
+
+#[derive(Clone)]
+pub struct LibP2pAdapter {
+    network_relay: OutboundRelay<<NetworkService<Libp2pNetworkBackend> as ServiceData>::Message>,
+    mixed_msgs: broadcast::Sender<Vec<u8>>,
+}
+
+#[async_trait::async_trait]
+impl NetworkAdapter for LibP2pAdapter {
+    type Backend = Libp2pNetworkBackend;
+
+    async fn new(
+        network_relay: OutboundRelay<<NetworkService<Self::Backend> as ServiceData>::Message>,
+    ) -> Self {
+        let relay = network_relay.clone();
+        let mixed_msgs = broadcast::Sender::new(BUFFER_SIZE);
+        let mixed_msgs_sender = mixed_msgs.clone();
+        // this wait seems to be helpful in some cases since we give the time
+        // to the network to establish connections before we start sending messages
+        tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+
+        tokio::spawn(async move {
+            let (sender, receiver) = tokio::sync::oneshot::channel();
+            if let Err((e, _)) = relay
+                .send(NetworkMsg::Subscribe {
+                    kind: Libp2pNetworkBackendEventKind::FullyMixedMessage,
+                    sender,
+                })
+                .await
+            {
+                tracing::error!("error subscribing to incoming mixed msgs: {e}");
+            }
+
+            let mut incoming_mixed_msgs = receiver.await.unwrap();
+            loop {
+                match incoming_mixed_msgs.recv().await {
+                    Ok(Libp2pNetworkBackendEvent::FullyMixedMessage(msg)) => {
+                        tracing::debug!("received a fully mixed message");
+                        if let Err(e) = mixed_msgs_sender.send(msg) {
+                            tracing::error!("error sending mixed message to consensus: {e}");
+                        }
+                    }
+                    Err(RecvError::Lagged(n)) => {
+                        tracing::error!("lagged messages: {n}")
+                    }
+                    Err(RecvError::Closed) => unreachable!(),
+                }
+            }
+        });
```

Comment:
I think this as a task is unnecessary?
When requesting for a subscriptrion you have a channel that can be transformed into a stream and map the necessary operations. This should be done on demand in the `mixed_msgs_stream` method directly.

Reply:
True. Fixed in 14b8de118e5eda5e8d1c3a20631e3ea6eff4fd84.
I adopted this style from the network adapter in the consensus service. If my fix looks good, I will update the network adapter in consensus as well in the next PR, to have the same style over our codebase.

Reply:
> True. Fixed in [14b8de1](https://github.com/logos-co/nomos-node/commit/14b8de118e5eda5e8d1c3a20631e3ea6eff4fd84). I adopted this style from the network adapter in the consensus service. But yes, we don't need to use the `tokio::sync::broadcast` channel redundantly here in the adapter level because the mix network backend is already using the broadcast channel. If my fix looks good, I will update the network adapter in consensus as well in the next PR, to have the same style over our codebase.

yep, this is it, much better like this.